gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/db/FabricDatabase.java

```diff
@@ -555,6 +565,181 @@ synchronized public void postCreate(
+//          SanityManager.DEBUG_PRINT("info", "cachedBatchTable = " + cachedBatchTable);
+          if (!internalColumnTablesSet.contains(cachedBatchTable)) {
+            tablesMissingColumnBuffer.add(storeTable);
+          }
+        }
+      }
+      if (!tablesMissingColumnBuffer.isEmpty()) {
+        SanityManager.DEBUG_PRINT("info",
+            "Catalog inconsistency detected: following column tables " +
+                "do not have column buffer: " +
+                "schema = " + storeEntry.getKey() + " tables = " + tablesMissingColumnBuffer);
+        dropTables(embedConn, storeEntry.getKey(), tablesMissingColumnBuffer);
+        removeTableFromHivestore(storeEntry.getKey(), tablesMissingColumnBuffer);
+      }
+    }
+  }
+
+  /**
+   * Remove Hive entries for which there is no DD entry
+   * @param hiveDBTablesMap schema to tables map of hive metastore entries
+   * @param gfDBTablesMap schema to tables map of DD entries
+   */
+  private static void removeInconsistentHiveEntries(
+      HashMap<String, List<String>> hiveDBTablesMap,
+      HashMap<String, List<String>> gfDBTablesMap) {
+    // remove tables that are in Hive store but not in datadictionary
+    for (Map.Entry<String, List<String>> hiveEntry : hiveDBTablesMap.entrySet()) {
+      List<String> storeTableList = gfDBTablesMap.get(hiveEntry.getKey());
+      List<String> hiveTableList = new LinkedList<>(hiveEntry.getValue());
+      if (!(storeTableList == null || storeTableList.isEmpty())) {
+        hiveTableList.removeAll(storeTableList);
+      }
+
+      if (!(hiveTableList == null || hiveTableList.isEmpty())) {
+        SanityManager.DEBUG_PRINT("info",
+            "Catalog inconsistency detected: following tables " +
+                "in Hive metastore are not in datadictionary: " +
+                "schema = " + hiveEntry.getKey() + " tables = " + hiveTableList);
+        removeTableFromHivestore(hiveEntry.getKey(), hiveTableList);
+      }
+    }
+  }
+
+  private static final void removeTableFromHivestore(String schema, List<String> tables) {
+    for (String table : tables) {
+      SanityManager.DEBUG_PRINT("info", "Removing table " +
+          schema + "." + table + " from Hive metastore");
+      Misc.getMemStoreBooting().getExternalCatalog().removeTable(schema, table, false);
+    }
+  }
+
+  private static final void dropTables(EmbedConnection embedConn,
+      String schema, List<String> tables) throws SQLException {
+    for (String table : tables) {
+      try {
+        SanityManager.DEBUG_PRINT("info", "Dropping table " +
+            schema + "." + table);
+
+        // drop column table
+        String cachedBatchTableName = com.gemstone.gemfire.
+            internal.snappy.CallbackFactoryProvider.getStoreCallbacks().
+            cachedBatchTableName(schema + "." + table);
+        embedConn.createStatement().execute(
+            "DROP TABLE IF EXISTS " + cachedBatchTableName);
+
+        // drop row buffer
```

Comment:
Check for the region size before dropping as it will suggest an inconsistent state.